File: src/lib/optimizer/improvementSuggestions.ts
import { API_ERROR_CODES, ApiError } from "@/lib/api/errors";
import type { AuthenticatedUser } from "@/lib/api/requireUser";
import type { AppDatabase } from "@/lib/db/client";
import { findAttachmentsBySetId, getSet } from "@/lib/db/repositories/setRepository";
import { listUserSets } from "@/lib/sets/setService";

import { optimizeFromSet } from "./optimizeFromSet";
import type { RankableCombination } from "./score";
import { parseOptimizerConstraints, type ArmorCombination } from "./types";

export type ImprovementSuggestion = {
  armorSetId: string;
  armorSetName: string;
  buildIds: string[];
  hasImprovement: boolean;
  betterCombination?: ArmorCombination;
  currentSummary?: RankableCombination;
};

export type ImprovementSuggestionsResponse = { suggestions: ImprovementSuggestion[] };

export type ImprovementSuggestionsInput = {
  db: AppDatabase;
  userId: number;
  auth: AuthenticatedUser;
  /** Hint that inventory just synced (attached constrained Sets). */
  afterSync?: boolean;
  /** On-open check for a single Set (attached or unattached). */
  armorSetId?: string;
};

/** Parse the improvement-suggestions query into a normalized input shape. */
export function parseImprovementSuggestionsQuery(url: URL): {
  afterSync: boolean;
  armorSetId?: string;
} {
  const armorSetId = url.searchParams.get("armorSetId");
  return {
    afterSync: url.searchParams.get("afterSync") === "1",
    ...(armorSetId ? { armorSetId } : {}),
  };
}

function attachedBuildIds(db: AppDatabase, setId: string): string[] {
  return [...new Set(findAttachmentsBySetId(db, setId).map((ref) => ref.buildId))];
}

async function suggestForSet(
  input: ImprovementSuggestionsInput,
  setId: string,
): Promise<ImprovementSuggestion> {
  const set = getSet(input.db, input.userId, setId);
  if (!set || set.type !== "armor") {
    throw new ApiError(API_ERROR_CODES.SET_NOT_FOUND, "Armor set not found", { setId }, 404);
  }
  const result = await optimizeFromSet({
    db: input.db,
    userId: input.userId,
    auth: input.auth,
    setId,
  });
  const top = result.combinations[0];
  return {
    armorSetId: setId,
    armorSetName: set.name,
    buildIds: attachedBuildIds(input.db, setId),
    hasImprovement: result.hasImprovement,
    ...(result.hasImprovement && top ? { betterCombination: top } : {}),
    ...(result.currentSummary ? { currentSummary: result.currentSummary } : {}),
  };
}

/**
 * Soft improvement suggestions (US5b): for a single opened Set, or for every
 * attached constrained Armor Set after a sync. Never mutates Set items — the
 * client confirms via apply-combination (suggest-then-confirm).
 */
export async function buildImprovementSuggestions(
  input: ImprovementSuggestionsInput,
): Promise<ImprovementSuggestionsResponse> {
  if (input.armorSetId) {
    return { suggestions: [await suggestForSet(input, input.armorSetId)] };
  }

  const suggestions: ImprovementSuggestion[] = [];
  for (const set of listUserSets(input.db, input.userId, { type: "armor" })) {
    if (!parseOptimizerConstraints(set.optimizerConstraints)) continue;
    if (attachedBuildIds(input.db, set.id).length === 0) continue;
    const suggestion = await suggestForSet(input, set.id);
    if (suggestion.hasImprovement) suggestions.push(suggestion);
  }
  return { suggestions };
}

File: flutter/packages/app/lib/src/improvement_suggestions.dart
/// Soft post-sync / on-open armor kit improvement suggestions (BR-OPT-004).
///
/// **Never mutates** sets — callers must confirm via [applyArmorCombinationInPlace].
library;

import 'package:destiny2_db/destiny2_db.dart' hide Build, SetItem, Synergy, SynergyLink;
import 'package:destiny2_domain/destiny2_domain.dart';

import 'optimizer_constraints_json.dart';
import 'set_use_cases.dart';

/// One soft suggestion for a constrained attached Armor Set.
class ImprovementSuggestion {
  const ImprovementSuggestion({
    required this.armorSetId,
    required this.armorSetName,
    required this.buildIds,
    required this.hasImprovement,
    this.betterCombination,
    this.currentSummary,
  });

  final String armorSetId;
  final String armorSetName;
  final List<String> buildIds;
  final bool hasImprovement;
  final ArmorCombination? betterCombination;
  final RankableCombination? currentSummary;
}

/// Injectable optimize for a single set (tests inject fixed response).
typedef OptimizeSetRunner = Future<ArmorOptimizeResponse> Function({
  required String setId,
  required ArmorSetOptimizerConstraints constraints,
  required List<CandidatePiece> candidates,
});

/// Build current rankable from set pieces matched against [candidates].
RankableCombination? currentRankableFromPieces({
  required List<SetItemRecord> activeItems,
  required List<CandidatePiece> candidates,
}) {
  final byInstance = {
    for (final c in candidates) c.instanceId: c,
  };
  final pieces = <CandidatePiece>[];
  for (final item in activeItems) {
    final id = item.instanceId;
    if (id == null || id.isEmpty) continue;
    final match = byInstance[id];
    if (match != null) pieces.add(match);
  }
  if (pieces.isEmpty) return null;
  final stats = estimateKitStats(pieces);
  final reuse = pieces.where((p) => p.usedInSets.isNotEmpty).length;
  return RankableCombination(
    estimatedStats: stats,
    reusePieceCount: reuse,
  );
}

/// Pure comparison gate used by suggestion builders.
bool combinationImprovesCurrent({
  required RankableCombination current,
  required ArmorCombination top,
  required ArmorSetOptimizerConstraints constraints,
}) {
  return detectImprovement(
    current,
    top.asRankable,
    constraints.statPriorities.isEmpty ? null : constraints.statPriorities,
    constraints.preferReuse,
  );
}

/// List constrained armor sets attached to ≥1 build (afterSync scan).
Future<List<SetDetail>> listConstrainedAttachedArmorSets(
  AppDatabase db,
  int userId,
) async {
  final armor = await listUserSets(db, userId, type: SetType.armor);
  final out = <SetDetail>[];
  for (final row in armor) {
    if (parseOptimizerConstraints(row.optimizerConstraints) == null) {
      continue;
    }
    final detail = await getSetDetail(db, userId, row.id);
    if (detail == null) continue;
    if (detail.usedBy.isEmpty) continue;
    out.add(detail);
  }
  return out;
}

/// Soft improvement suggestions — **never writes**.
///
/// When [afterSync] is true (default), scans all constrained attached armor
/// sets. When [armorSetId] is set, evaluates that set only (attached or not).
Future<List<ImprovementSuggestion>> buildImprovementSuggestions(
  AppDatabase db,
  int userId, {
  required List<CandidatePiece> candidates,
  OptimizeSetRunner? optimizeRunner,
  bool afterSync = true,
  String? armorSetId,
  bool hasInventory = true,
}) async {
  final runner = optimizeRunner ??
      ({
        required String setId,
        required ArmorSetOptimizerConstraints constraints,
        required List<CandidatePiece> candidates,
      }) async {
        return optimizeArmorCore(
          ArmorOptimizeRequest(
            candidates: candidates,
            constraints: KitConstraints(
              lockedExoticItemHash: constraints.lockedExoticItemHash,
              requireExotic: constraints.requireExotic,
              setBonusGoals: constraints.setBonusGoals,
            ),
            statPriorities: constraints.statPriorities,
            statThresholds: constraints.statThresholds,
            requireThresholds: constraints.requireThresholds,
            preferReuse: constraints.preferReuse,
            maxResults: 5,
            hasInventory: hasInventory,
          ),
        );
      };

  Future<ImprovementSuggestion?> suggestFor(SetDetail detail) async {
    final constraints =
        parseOptimizerConstraints(detail.set.optimizerConstraints);
    if (constraints == null) return null;

    final buildIds = <String>{
      for (final ref in detail.usedBy) ref.buildId,
    }.toList()
      ..sort();

    final response = await runner(
      setId: detail.set.id,
      constraints: constraints,
      candidates: candidates,
    );
    final top = response.combinations.isEmpty
        ? null
        : response.combinations.first;
    final current = currentRankableFromPieces(
      activeItems: detail.activeItems,
      candidates: candidates,
    );

    final hasImprovement = top != null &&
        current != null &&
        combinationImprovesCurrent(
          current: current,
          top: top,
          constraints: constraints,
        );

    return ImprovementSuggestion(
      armorSetId: detail.set.id,
      armorSetName: detail.set.name,
      buildIds: buildIds,
      hasImprovement: hasImprovement,
      betterCombination: hasImprovement ? top : null,
      currentSummary: current,
    );
  }

  if (armorSetId != null) {
    final detail = await getSetDetail(db, userId, armorSetId);
    if (detail == null || detail.set.type != SetType.armor.wireName) {
      return const [];
    }
    final s = await suggestFor(detail);
    return s == null ? const [] : [s];
  }

  if (!afterSync) return const [];

  final targets = await listConstrainedAttachedArmorSets(db, userId);
  final suggestions = <ImprovementSuggestion>[];
  for (final detail in targets) {
    final s = await suggestFor(detail);
    if (s != null && s.hasImprovement) {
      suggestions.add(s);
    }
  }
  return suggestions;
}
